airesources/OCaml/game.ml

```diff
@@ -0,0 +1,251 @@
+(* OCaml Starter for Halite II on Halite.io
+   This code is public domain. There is no warranty.
+ *)
+
+open Td;;
+
+let set_planets_docked_ships state =
+  List.iter (fun planet ->
+    planet.docked_ships <- List.filter (fun ship ->
+      ship.status = `Docked
+      && ship.docked_planet_id = planet.p_entity.id
+    ) state.all_ships
+  ) state.planets
+;;
+
+let update_map state =
+  state.turn <- state.turn + 1;
+  state.last_update <- Unix.gettimeofday ();
+  Debug.info (Printf.sprintf "--- Begin Turn %i ---\n" state.turn);
+  Networking.parse_map state (Networking.get_string());
+  set_planets_docked_ships state;
+;;
+
+let init state name initial_bot_setup =
+  state.my_id <- int_of_string (Networking.get_string ());
+  Debug.setup_logging state.my_id name;
+  Networking.read_width_height state (Networking.get_string ());
+  update_map state;
+  initial_bot_setup state;
+  Networking.send_string name;
```

Comment:
That's not going to work I think for #151 - the 60 second initial timeout ends as soon as you send the bot name over. This has to be done right before first map update.

Reply:
I seem to be misunderstanding something then... is there anything useful the bot can do with those 60 seconds if it hasn't seen the map state yet?

This code does appear to work with the distributed Halite engine - if I output the state of the planets in that `initial_bot_setup` function, it does have the complete list of planets at that point.

In the Python starter, the `__init__` function in networking.py calls `update_map` with `_send_name = False` before setting `_send_name = True" - to me this seems to have the same result as the OCaml code here, but I'm guessing I've missed something. I'll await clarification.

Reply:
It's possible I misunderstood something. Here is protocol:

- engine sends initial map to bot kit and starts the 60 second timer
- bot kit framework receives initial map and shows it to the bot
- bot author can do whatever it wants with the initial map for 60 seconds, such as prebuild pathfinding-related stuff
- bot author notifies bot kit framework that they are done, either explicitly or asking for first map update
- bot kit framework sends bot name to game engine, thus terminating the 60 second timer
- game engine sends first turn map over - actually identical to initial map - and the normal 2 second turn timer starts

Does this help?

Reply:
Actually, how about you update MyBot to show exactly where the user can take the initial map and spend 60 seconds analyzing it? Sorta like what cpp and Java kits do - take a look at their MyBot examples.

Reply:
That `initial_bot_setup` function in where the user would do their 60 seconds of processing. It's a blank function in MyBot.ml with the comment "Anything you want to do with your opening time".

Reply:
Regarding .mli files, I'll look into it. I don't normally use them at all, but I'll see if I can do that. It'll take some time.